File: python/Permutations/Permutations.py

class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        n = len(nums)
        res=[]
        if n == 0:
            return res
        out = []
        visited = [0 for i in range(n)]
        self.DFS(nums,0,visited,out,res)
        return res
    def DFS(self, nums: List[int], level : int, visited: List[int], out: List[int], res: List[List[int]]):
        n = len(nums)
        if level == n:
            res.append(out[:])
            return
        for i in range (0,n):
            if visited[i] == 1:
                continue
            visited[i] = 1
            out.append(nums[i])
            self.DFS(nums, level+1, visited, out, res)
            out.pop()  ## del out[-1] out.pop(-1)
            visited[i] = 0

File: cpp/Permutations/Permutations.cpp

class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> res;
        if (nums.empty()) return res;
        vector<int> out;
        int n =nums.size();
        vector<int> visited(n,0);
        DFS(nums,0,visited,out,res);
        return res;
    }
    
    void DFS(vector<int>& nums, int level, vector<int>& visited, vector<int>& out, vector<vector<int>>& res)
    {
        if (level == nums.size())
        {
            res.push_back(out);
            return;
        }
        for (int i=0; i<nums.size();i++)
        {
            if (visited[i] == 1)
                continue;
            visited[i] = 1;
            out.push_back(nums[i]);
            DFS(nums,level+1,visited,out,res);
            out.pop_back();
            visited[i] = 0;
        }        
    }
};
